client/network/sync/src/warp.rs

```diff
@@ -70,21 +80,58 @@ where
 	B: BlockT,
 	Client: HeaderBackend<B> + ProofProvider<B> + 'static,
 {
-	///  Create a new instance.
-	pub fn new(client: Arc<Client>, warp_sync_provider: Arc<dyn WarpSyncProvider<B>>) -> Self {
+	/// Create a new instance. When passing a warp sync provider we will be checking for proof and
+	/// authorities. Alternatively we can pass a target block when we want to skip downloading
+	/// proofs, in this case we will continue polling until the target block is known.
+	pub fn new(client: Arc<Client>, warp_sync_params: WarpSyncParams<B>) -> Self {
 		let last_hash = client.hash(Zero::zero()).unwrap().expect("Genesis header always exists");
-		let phase = Phase::WarpProof {
-			set_id: 0,
-			authorities: warp_sync_provider.current_authorities(),
-			last_hash,
+		match warp_sync_params {
+			WarpSyncParams::WithProvider(warp_sync_provider) => {
+				let phase = Phase::WarpProof {
+					set_id: 0,
+					authorities: warp_sync_provider.current_authorities(),
+					last_hash,
+					warp_sync_provider: warp_sync_provider.clone(),
+				};
+				Self { client, phase, total_proof_bytes: 0 }
+			},
+			WarpSyncParams::WaitForTarget(block) => Self {
+				client,
+				phase: Phase::PendingTargetBlock { target_block: block },
+				total_proof_bytes: 0,
+			},
+		}
+	}
+
+	/// Poll to make progress.
+	///
+	/// This only makes progress when `phase = Phase::PendingTargetBlock` and the pending block was
+	/// send.
+	pub fn poll(&mut self, cx: &mut std::task::Context) -> Poll<()> {
+		let new_phase = if let Phase::PendingTargetBlock { target_block } = &mut self.phase {
+			match target_block.poll_unpin(cx) {
+				Poll::Ready(Ok(target_block)) => Some(Phase::TargetBlock(target_block)),
+				Poll::Ready(Err(e)) => {
+					error!(target: "sync", "Failed to get target block. Error: {:?}",e);
+					None
+				},
+				_ => None,
+			}
+		} else {
+			None
 		};
-		Self { client, warp_sync_provider, phase, total_proof_bytes: 0 }
+
+		if let Some(new_phase) = new_phase {
+			self.phase = new_phase;
+		}
+
+		Poll::Pending
```

Comment:
Small nit, should only receive Pending when one of the inner poll do return Pending (given it is a oneshot and how the function is called, it is not actually an issue).

Reply:
In the context of the Poll, we are pending until the target block is reached right? So surely Pending is the correct thing to return here?

Reply:
no issue the current way it is use, but Poll::Pending should only be use when something can wake the thread up.
I would just return a bool that indicate if we can drop the future later.
(got a bad time recently with some of my code that did hang because of Pending being return (and propagated) but nothing could wake up the thread).
But that is just a minor nit, you can ignore it :)

Reply:
@cheme is a little bit right here. We can just change the entire function to not return `Poll` at all, because the outer "contract" is any way that it wakes it up automatically for other events.